File: LegacySource/settlement/room/health/hospital/Bed.cs
using game.audio;
using game.faction;
using init.resources;
using settlement.entity.humanoid;
using settlement.main;
using settlement.misc.job;
using settlement.misc.util;
using settlement.room.main;
using settlement.room.main.util;
using snake2d.util.datatypes;

namespace settlement.room.health.hospital
{
    public class Bed
    {
        private readonly Coo coo = new Coo();
        private HospitalInstance ins;

        private readonly RoomBits wreserved = new RoomBits(coo, 0b0000_0000_0001);
        private readonly RoomBits wres1 = new RoomBits(coo, 0b0000_0000_1110);
        private readonly RoomBits wres2 = new RoomBits(coo, 0b0000_1111_0000);
        private readonly RoomBits freeWork = new RoomBits(coo, 0b0000_1000_0000);
        private readonly RoomBits sstate = new RoomBits(coo, 0b1111_0000_0000)
        {
            public override void Set(ROOMA r, int t)
            {
                ins.service().Report(service, ins.blueprintI().service(), -1);
                base.Set(r, t);
                ins.service().Report(service, ins.blueprintI().service(), 1);
            }
        };

        private const int I_UNAVAILABLE = 0;
        private const int I_AVAILABLE = 1;
        private const int I_RESERVED = 2;
        private const int I_USED = 3;

        private static readonly Bed self = new Bed();

        public static SETT_JOB Job(int tx, int ty)
        {
            if (self.Init(tx, ty))
                return self.job;
            return null;
        }

        public static FSERVICE Service(int tx, int ty)
        {
            if (self.Init(tx, ty))
                return self.service;
            return null;
        }

        public static bool Res1(int tx, int ty)
        {
            if (!self.Init(tx, ty))
                return false;
            return self.wres1.Get() > 0;
        }

        public static bool Res2(int tx, int ty)
        {
            if (!self.Init(tx, ty))
                return false;
            return self.wres2.Get() > 0;
        }

        public static void Consume(int tx, int ty)
        {
        }

        private bool Init(int tx, int ty)
        {
            ins = B().Get(tx, ty);
            if (ins != null && SETT.ROOMS().FData.TileData.Get(tx, ty) == Constructor.CODE_S)
            {
                coo.Set(tx, ty);
                return true;
            }
            return false;
        }

        public static bool Made(int tx, int ty)
        {
            return self.Init(tx, ty) && self.sstate.Get() != self.I_UNAVAILABLE;
        }

        public static bool Resource(int tx, int ty)
        {
            return self.Init(tx, ty) && self.wres1.Get() > 0 && self.sstate.Get() != self.I_USED;
        }

        private static ROOM_HOSPITAL B()
        {
            return SETT.ROOMS().HOSPITAL;
        }

        private readonly SETT_JOB job = new SETT_JOB
        {
            private readonly int wt = 30;
            private readonly RBITImp bits = new RBITImp();

            public override bool JobUseTool()
            {
                return false;
            }

            public override bool JobUseHands()
            {
                return sstate.Get() == I_UNAVAILABLE;
            }

            public override void JobStartPerforming()
            {
            }

            public override SoundRace JobSound()
            {
                return B().Employment().Sound();
            }

            public override RBIT JobResourceBitToFetch()
            {
                bits.Clear();
                if (sstate.Get() == I_UNAVAILABLE)
                    return null;
                if (wres1.Get() == 0 && ins.Fetch[0] && B().ResLocks.Get(0).Passes(FACTIONS.player()))
                    bits.Or(B().Indus.Get(0).Ins().Get(0).Resource);
                if (wres2.Get() == 0 && ins.Fetch[1] && B().ResLocks.Get(1).Passes(FACTIONS.player()))
                    bits.Or(B().Indus.Get(0).Ins().Get(1).Resource);
                return bits.IsClear() ? null : bits;
            }

            public override bool JobReservedIs(RESOURCE r)
            {
                return wreserved.Get() == 1;
            }

            public override void JobReserveCancel(RESOURCE r)
            {
                wreserved.Set(ins, 0);
            }

            public override bool JobReserveCanBe()
            {
                if (wreserved.Get() == 1)
                    return false;
                if (sstate.Get() == I_UNAVAILABLE)
                    return true;
                if (wres1.Get() == 0 && ins.Fetch[0] && B().ResLocks.Get(0).Passes(FACTIONS.player()))
                    return true;
                if (wres2.Get() == 0 && ins.Fetch[1] && B().ResLocks.Get(1).Passes(FACTIONS.player()))
                    return true;
                return false;
            }

            public override void JobReserve(RESOURCE r)
            {
                wreserved.Set(ins, 1);
            }

            public override double JobPerformTime(Humanoid a)
            {
                return freeWork.Get() == 1 ? 1 : wt;
            }

            public override RESOURCE JobPerform(Humanoid skill, RESOURCE r, int rAm)
            {
                JobReserveCancel(r);
                if (r == B().Indus.Get(0).Ins().Get(0).Resource)
                {
                    B().Indus.Get(0).Ins().Get(0).Inc(ins, rAm);
                    wres1.Inc(ins, rAm);
                }
                else if (r == B().Indus.Get(0).Ins().Get(1).Resource)
                {
                    B().Indus.Get(0).Ins().Get(1).Inc(ins, rAm);
                    wres2.Inc(ins, rAm);
                }
                else
                {
                    if (sstate.Get() == I_UNAVAILABLE)
                    {
                        sstate.Set(ins, I_AVAILABLE);
                    }
                }
                freeWork.Set(ins, ins.Employees().FetchBonusConsume(wt + 1) ? 1 : 0);
                return null;
            }

            public override CharSequence JobName()
            {
                return B().Employment().Verb;
            }

            public override COORDINATE JobCoo()
            {
                return coo;
            }
        };

        public readonly FSERVICE service = new FSERVICE
        {
            public override int Y()
            {
                return coo.Y();
            }

            public override int X()
            {
                return coo.X();
            }

            public override bool FindableReservedIs()
            {
                return sstate.Get() == I_RESERVED || sstate.Get() == I_USED;
            }

            public override bool FindableReservedCanBe()
            {
                return sstate.Get() == I_AVAILABLE;
            }

            public override void FindableReserveCancel()
            {
                sstate.Set(ins, I_AVAILABLE);
            }

            public override void FindableReserve()
            {
                if (FindableReservedCanBe())
                    sstate.Set(ins, I_RESERVED);
            }

            public override void StartUsing()
            {
                sstate.Set(ins, I_USED);
            }

            public override void Consume()
            {
                wres1.Inc(ins, -1);
                wres2.Inc(ins, -1);
                sstate.Set(ins, I_UNAVAILABLE);
                ins.Jobs.SearchAgain();
            }
        };
    }
}

File: settlement/room/health/hospital/Bed.java
package settlement.room.health.hospital;

import game.audio.SoundRace;
import game.faction.FACTIONS;
import init.resources.RBIT;
import init.resources.RBIT.RBITImp;
import init.resources.RESOURCE;
import settlement.entity.humanoid.Humanoid;
import settlement.main.SETT;
import settlement.misc.job.SETT_JOB;
import settlement.misc.util.FSERVICE;
import settlement.room.main.ROOMA;
import settlement.room.main.util.RoomBits;
import snake2d.util.datatypes.COORDINATE;
import snake2d.util.datatypes.Coo;

final class Bed {

	private final Coo coo = new Coo();
	private HospitalInstance ins;
	
	private final RoomBits wreserved = 		new RoomBits(coo, 0b0000_0000_0001);
	private final RoomBits wres1 = 			new RoomBits(coo, 0b0000_0000_1110);
	private final RoomBits wres2 = 			new RoomBits(coo, 0b0000_0111_0000);
	private final RoomBits freeWork = 		new RoomBits(coo, 0b0000_1000_0000);
	private final RoomBits sstate = 		new RoomBits(coo, 0b1111_0000_0000) {
		
		@Override
		public void set(ROOMA r, int t) {
			ins.service().report(service, ins.blueprintI().service(), -1);
			super.set(r, t);
			ins.service().report(service, ins.blueprintI().service(), 1);
		};
		
	};
	
	private final int I_UNAVAILABLE = 0;
	private final int I_AVAILABLE = 1;
	private final int I_RESERVED = 2;
	private final int I_USED = 3;
	
	private static Bed self = new Bed();
	
	public static SETT_JOB job(int tx, int ty) {
		if (self.init(tx, ty))
			return self.job;
		return null;
	}
	
	public static FSERVICE service(int tx, int ty) {
		if (self.init(tx, ty))
			return self.service;
		return null;
	}
	public static boolean res1(int tx, int ty) {
		if (!self.init(tx, ty))
			return false;
		return self.wres1.get() > 0;
	}
	
	public static boolean res2(int tx, int ty) {
		if (!self.init(tx, ty))
			return false;
		return self.wres2.get() > 0;
	}
	
	public static void consume(int tx, int ty) {
		
	}
	
	private boolean init(int tx, int ty) {
		ins = b().get(tx, ty);
		if (ins != null && SETT.ROOMS().fData.tileData.get(tx, ty) == Constructor.CODE_S) {
			coo.set(tx, ty);
			return true;
		}
		return false;
	}
	
	public static boolean made(int tx, int ty) {
		return self.init(tx, ty) && self.sstate.get() != self.I_UNAVAILABLE;
	}
	
	public static boolean resource(int tx, int ty) {
		return self.init(tx, ty) && self.wres1.get() > 0 && self.sstate.get() != self.I_USED;
	}
	
	private static ROOM_HOSPITAL b() {
		return SETT.ROOMS().HOSPITAL;
	}
	
	private final SETT_JOB job = new SETT_JOB() {
		
		private final int wt = 30;
		private final RBITImp bits = new RBITImp();
		
		@Override
		public boolean jobUseTool() {
			return false;
		}
		
		@Override
		public boolean jobUseHands() {
			return sstate.get() == I_UNAVAILABLE;
		};
		
		@Override
		public void jobStartPerforming() {
			
		}
		
		@Override
		public SoundRace jobSound() {
			return b().employment().sound();
		}
		
		@Override
		public RBIT jobResourceBitToFetch() {
			bits.clear();
			if (sstate.get() == I_UNAVAILABLE)
				return null;
			if (wres1.get() == 0 && ins.fetch[0] && b().resLocks.get(0).passes(FACTIONS.player()))
				bits.or(b().indus.get(0).ins().get(0).resource);
			if (wres2.get() == 0 && ins.fetch[1] && b().resLocks.get(1).passes(FACTIONS.player()))
				bits.or(b().indus.get(0).ins().get(1).resource);
			return bits.isClear() ? null : bits;
		}
		
		@Override
		public boolean jobReservedIs(RESOURCE r) {
			return wreserved.get() == 1;
		}
		
		@Override
		public void jobReserveCancel(RESOURCE r) {
			wreserved.set(ins, 0);
		}
		
		@Override
		public boolean jobReserveCanBe() {
			if (wreserved.get() == 1)
				return false;
			if (sstate.get() == I_UNAVAILABLE)
				return true;
			if (wres1.get() == 0 && ins.fetch[0] && b().resLocks.get(0).passes(FACTIONS.player()))
				return true;
			if (wres2.get() == 0 && ins.fetch[1] && b().resLocks.get(1).passes(FACTIONS.player()))
				return true;
			return false;
			
		}
		
		@Override
		public void jobReserve(RESOURCE r) {
			wreserved.set(ins, 1);
		}
		
		@Override
		public double jobPerformTime(Humanoid a) {
			return freeWork.get() == 1 ? 1 : wt;
		}
		
		@Override
		public RESOURCE jobPerform(Humanoid skill, RESOURCE r, int rAm) {
			jobReserveCancel(r);
			if (r == b().indus.get(0).ins().get(0).resource) {
				b().indus.get(0).ins().get(0).inc(ins, rAm);
				wres1.inc(ins, rAm);
			}else if (r == b().indus.get(0).ins().get(1).resource) {
				b().indus.get(0).ins().get(1).inc(ins, rAm);
				wres2.inc(ins, rAm);
			}else {
				if (sstate.get() == I_UNAVAILABLE) {
					sstate.set(ins, I_AVAILABLE);
				}
			}
			freeWork.set(ins, ins.employees().fetchBonusConsume(wt+1) ? 1 : 0);
			return null;
		}
		
		@Override
		public CharSequence jobName() {
			return b().employment().verb;
		}
		
		@Override
		public COORDINATE jobCoo() {
			return coo;
		}
		
		
	};
	
	public final FSERVICE service = new FSERVICE() {
		
		@Override
		public int y() {
			return coo.y();
		}
		
		@Override
		public int x() {
			return coo.x();
		}
		
		@Override
		public boolean findableReservedIs() {
			return sstate.get() == I_RESERVED || sstate.get() == I_USED;
		}
		
		@Override
		public boolean findableReservedCanBe() {
			return sstate.get() == I_AVAILABLE;
		}
		
		@Override
		public void findableReserveCancel() {
			sstate.set(ins, I_AVAILABLE);
		}
		
		@Override
		public void findableReserve() {
			if (findableReservedCanBe())
				sstate.set(ins, I_RESERVED);
		}
		
		@Override
		public void startUsing() {
			sstate.set(ins, I_USED);
		};
		
		@Override
		public void consume() {
			wres1.inc(ins, -1);
			wres2.inc(ins, -1);
			sstate.set(ins, I_UNAVAILABLE);
			ins.jobs.searchAgain();
		}
	};
	
}
